### PR TITLE
[SPARK-44503][SQL] Add analysis and planning for PARTITION BY and ORDER BY clause after TABLE arguments for TVF calls

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds analysis and planning for PARTITION BY and ORDER BY clause after TABLE arguments for TVF calls.

* This PR appends `RepartitionByExpressions` and `Sort` logical operators as needed for this purpose.
* This PR also adds `WITH SINGLE PARTITION` syntax to `TABLE` arguments. This indicates that exactly one instance of the UDTF should process all rows in the input relation, and is mutually exclusive with the `PARTITION BY` clause. Then we append `Repartition` instead of `RepartitionBy` to implement this.
* In addition to up dated parsing unit tests, this PR adds end-to-end testing in the Scala `PythonUDTFSuite`. Note that full query execution is not implemented yet, that will come in the next PR since further execution support is needed to detect when partition boundaries end by inspecting the values of keys in consecutive rows.

### Why are the changes needed?

This will provide a way for the TVF caller to indicate desired semantics for dividing up the rows of the input table into partitions for consumption by the underlying algorithm.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR adds unit test coverage and also end-to-end tests in Python.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @allisonwang-db @ueshin @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -2073,6 +2073,13 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
               _.containsPattern(FUNCTION_TABLE_RELATION_ARGUMENT_EXPRESSION), ruleId)  {
               case t: FunctionTableSubqueryArgumentExpression =>
                 val alias = SubqueryAlias.generateSubqueryName(s"_${tableArgs.size}")
+                resolvedFunc match {
+                  case Generate(_: PythonUDTF, _, _, _, _, _) =>
+                  case _ if t.hasRepartitioning =>
+                    throw QueryCompilationErrors.tableValuedFunctionPartitionByClauseNotSupported(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2079]: Can we add a test for this case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala [hunk 2]
```diff
@@ -32,11 +32,22 @@ import org.apache.spark.sql.types.DataType
...
  * a table subquery that may itself refer to one or more tables in its own
  * FROM clause.
+ *
+ * Each TABLE argument may also optionally include a PARTITION BY clause. If present, these indicate
+ * how to logically split up the input relation such that the table-valued function evaluates
+ * exactly once for each partition, and returns the union of all results. If no partitioning list is
+ * present, this splitting of the input relation is undefined. Furthermore, if the PARTITION BY
+ * clause includes a following ORDER BY clause, Catalyst will sort the rows in each partition such
+ * that the table-valued function receives them one-by-one in the requested order. Otherwise, if no
+ * such ordering is specified, the ordering of rows within each partition is undefined.
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala:42]: Maybe we can use the javadoc style here to explain each parameter? `@param`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala [hunk 3]
```diff
@@ -61,5 +75,36 @@ case class FunctionTableSubqueryArgumentExpression(
...
 
-  lazy val evaluable: LogicalPlan = Project(Seq(Alias(CreateStruct(plan.output), "c")()), plan)
+  def hasRepartitioning: Boolean = withSinglePartition || partitionByExpressions.nonEmpty
+
+  lazy val evaluable: LogicalPlan = {
+    val subquery = if (hasRepartitioning) {
+      // If the TABLE argument includes the WITH SINGLE PARTITION or PARTITION BY or ORDER BY
+      // clause(s), add a corresponding logical operator to represent the repartitioning operation
+      // in the query plan.
+      RepartitionForTableFunctionCall(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala:85]: Just curious, have we considered reusing `Sort` + `RepartitionByExpression`/`Repartition` nodes, instead of having a dedicated logical plan?

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala]: Good idea, I updated the code to use this, now we don't need to add a new operator!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala:42]: Sounds good, done.

File: python/pyspark/sql/tests/test_udtf.py [hunk 4]
```diff
@@ -1444,6 +1447,83 @@ def terminate(self):
...
+                r'plan_id=[\d]+', r'plan_id=xx', stripExprIds)
+            stripEvalType = re.sub(
+                r'\+- .....EvalPythonUDTF test_udtf_pb.*', r'+- EvalPythonUDTF test_udtf_pb',
+                stripPlanIds)
+            print('Query plan: ' + stripEvalType)
+            return stripEvalType.strip('\n')
+        def expected(input: str) -> str:
+            return textwrap.dedent(input).strip('\n')
+
+        self.assertEqual(
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/sql/tests/test_udtf.py]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2079]: Python UDTFs are the only type of table function that support input relation arguments right now IIRC in Spark. Certain Spark forks may define their own and then this check would apply. (Please let me know if you can think of any way to test this in this PR.)

**Comment by dtenedor** [comment]
Hi @allisonwang-db  thanks for your review, I have responded to your comments, please take another look 🙏

**Comment by allisonwang-db** [comment]
@dtenedor since the TABLE argument support is already in Spark 3.5, should we add this to Spark 3.5 as well? also cc @cloud-fan

File: python/pyspark/sql/tests/test_udtf.py [hunk 5]
```diff
@@ -14,9 +14,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+
 import os
+import re
 import shutil
 import tempfile
+import textwrap
```
Comment (cloud-fan) [inline python/pyspark/sql/tests/test_udtf.py:22]: unnecessary change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2079]: can we test the builtin TVFs like `explode`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -2445,8 +2452,12 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
           InSubquery(values, expr.asInstanceOf[ListQuery])
         case s @ LateralSubquery(sub, _, exprId, _, _) if !sub.resolved =>
           resolveSubQuery(s, outer)(LateralSubquery(_, _, exprId))
-        case a @ FunctionTableSubqueryArgumentExpression(sub, _, exprId) if !sub.resolved =>
-          resolveSubQuery(a, outer)(FunctionTableSubqueryArgumentExpression(_, _, exprId))
+        case a @ FunctionTableSubqueryArgumentExpression(
+            sub, _, exprId, partitionByExpressions, withSinglePartition, orderByExpressions)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2456]: nit:
```
case a: FunctionTableSubqueryArgumentExpression if !a. plan.resolved =>
  resolveSubQuery(a, outer)((plan, outerAttrs) => a.copy(plan = plan, outerAttrs = outerAttrs))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala [hunk 7]
```diff
@@ -32,11 +32,40 @@ import org.apache.spark.sql.types.DataType
...
  */
 case class FunctionTableSubqueryArgumentExpression(
     plan: LogicalPlan,
     outerAttrs: Seq[Expression] = Seq.empty,
-    exprId: ExprId = NamedExpression.newExprId)
+    exprId: ExprId = NamedExpression.newExprId,
+    partitionByExpressions: Seq[Expression] = Seq.empty,
+    withSinglePartition: Boolean = false,
+    orderByExpressions: Seq[SortOrder] = Seq.empty)
   extends SubqueryExpression(plan, outerAttrs, exprId, Seq.empty, None) with Unevaluable {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/FunctionTableSubqueryArgumentExpression.scala:70]: shall we add an assert here that `partitionByExpressions` must be Nil if `withSinglePartition` is true?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 8]
```diff
@@ -1564,14 +1564,43 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
...
+        }
+      }
+    Option(ctx.tableArgumentPartitioning)
+      .map(_.partition.asScala.map(expression))
+      .foreach { expressions =>
+        if (expressions.nonEmpty) {
+          partitionByExpressions = expressions
+        }
+      }
+    Option(ctx.tableArgumentPartitioning)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1583]: nit:
```
Option(ctx.tableArgumentPartitioning).foreah { p =>
  if (p.SINGLE != null) {
    withSinglePartition = true
  }
  partitionByExpressions = p.partition.asScala.map(expression)
  orderByExpressions = p.sortItem.asScala.map(visitSortItem)
}
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
let's backport to make the feature complete. cc release manager @xuanyuanking 

@dtenedor can you update the PR description? It's outdated as we don't add new query plans at the end.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 8 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Thanks, this is better.

**Review by dtenedor — COMMENTED** [review]
Thanks @cloud-fan  for your review, made suggested changes!

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master/3.5.

**Comment by ueshin** [comment]
@dtenedor There were some conflicts with `branch-3.5`. Could you help fix and submit another PR against `branch-3.5`? Thanks.

**Comment by ueshin** [comment]
@dtenedor @allisonwang-db @cloud-fan 
Oh, wait. the previous PR about the grammar for PARTITION BY and ORDER BY clause was merged only in `master`.
If we need to support this in 3.5, could you also submit a PR to backport it, too?

**Comment by dtenedor** [comment]
Discussed offline, we decided not to backport anymore PRs from this feature to Spark 3.5.
